libbeat/cmd/instance/beat.go

```diff
@@ -1466,3 +1476,49 @@ func sanitizeIPs(ips []string) []string {
 	}
 	return validIPs
 }
+
+func promoteOutputQueueSettings(bc *beatConfig) error {
+	if bc.Output.IsSet() && bc.Output.Config().Enabled() {
+		pc := pipeline.Config{}
+		err := bc.Output.Config().Unpack(&pc)
+		if err != nil {
+			return fmt.Errorf("error unpacking output queue settings: %w", err)
+		}
+		if pc.Queue.IsSet() {
+			logp.Info("global queue settings replaced with output queue settings")
+			bc.Pipeline.Queue = pc.Queue
+		}
+	}
+	return nil
+}
+
+func (bc *beatConfig) Validate() error {
+	if bc.Output.IsSet() && bc.Output.Config().Enabled() {
+		pc := pipeline.Config{}
+		err := bc.Output.Config().Unpack(&pc)
+		if err != nil {
+			return fmt.Errorf("error unpacking output queue settings: %w", err)
+		}
+		if bc.Pipeline.Queue.IsSet() && pc.Queue.IsSet() {
+			return fmt.Errorf("top level queue and output level queue settings defined, only one is allowed")
+		}
+	}
+	return nil
+}
+
+// checkAgentDiskQueue should be run after management.NewManager() so
+// that publisher.UnderAgent will be set with correct value
+func checkAgentDiskQueue(bc *beatConfig) error {
```

Comment:
I tested this and it doesn't seem to work:

```sh
sudo elastic-agent inspect
...
outputs:
  default:
    api_key: sfga4ooBeuYcqqQuT4_b:h4h_6lRaTEqN48FSOwST-A
    hosts:
    - https://60a01e9179764ca0b9c2e2fbf47d6d67.eastus2.staging.azure.foundit.no:443
    queue:
      disk:
        max_size: 10GB
    type: elasticsearch
path:
  config: /Library/Elastic/Agent
  data: /Library/Elastic/Agent/data
  home: /Library/Elastic/Agent/data/elastic-agent-123ba9
  logs: /Library/Elastic/Agent
sudo elastic-agent status
┌─ fleet
│  └─ status: (HEALTHY) Connected
└─ elastic-agent
   └─ status: (HEALTHY) Running
```

Steps to reproduce:

1. Build metricbeat and filebeat from this branch. I used `cd x-pack/metricbeat && mage build`. 
2. Build or download an 8.11 snapshot agent.
3. Copy the built metricbeat and filebeat into the `data/elastic-agent-$hash/components` directory of the agent. `cp ~/go/src/github.com/elastic/beats/x-pack/metricbeat/metricbeat /Users/cmackenzie/Downloads/elastic-agent-8.11.0-SNAPSHOT-darwin-aarch64
/data/elastic-agent-123ba9/components`
4. Install the agent. I enrolled it with Fleet but you could test this with a standalone agent too.
5. Configure the disk queue in the output as shown above.
6. Expect to see an error but see nothing.

I also saw no obvious logs about the queue type being configured in the logs or metrics. I suspect the memory queue configuration might be getting ignored as well.

I did confirm the running version of the beat was the most recent commit from this branch.

Reply:
For Fleet I had the following in the advanced yaml parameters box of the output, with the system integration installed with logs and monitoring enabled.

```yaml
queue.disk:
  max_size: 10GB
```

Reply:
This is what I see in the `beat-rendered-config.yml` for the `log-default` component:

```
outputs:
    elasticsearch:
        api_key: <REDACTED>
        bulk_max_size: 50
        hosts:
            - https://<REDACTED>.eastus2.staging.azure.foundit.no:443
        queue:
            disk:
                max_size: 10GB
        type: elasticsearch
```

Reply:
Something on the agent output reload path might not be reloading the queue settings:

https://github.com/elastic/beats/blob/61c2d282416a7054de4eb2411d51dfb8a5fc6859/x-pack/libbeat/management/managerV2.go#L689-L726

Reply:
I changed it to be a Validation check, which should happen whenever the config is unpacked.  So that should fix the disk queue problem.

The reload may be more interesting.  I think this will mean every output change requires a full restart since you need the queue settings earlier to setup the pipeline.

Reply:
We already restart the Beats when the output configuration changes, it looks optional in the code but it is enabled in all the Beat spec files: https://github.com/elastic/elastic-agent/blob/123ba9ce80c9865f72fa3659b5cafe9b51954f49/specs/filebeat.spec.yml#L32-L33

```yaml
        - "-E"
        - "management.restart_on_output_change=true"
```

We also need restarts whenever any of the TLS configuration changes for example.